data-processing/common/parse_equation.py

```diff
@@ -1,151 +1,290 @@
-            index = _get_character_index(node)
-            if index is not None:
-                characters.append(index)
-
-        # Group consecutive character symbols only if they are a child of the mrow tag.
-        # The .has_attr('s2:end') is necessary as not every symbol is annotated by katex
-        if symbol_node.has_attr('s2:end') and symbol_node.parent.name in MERGABLE_PARENT_TAGS and symbol_node.name in CHARACTER_TAGS:
-            # This symbol_node is part of consecutive_char_sequence but not the current sequence.
-            if len(consecutive_char_sequence) > 0 and not consecutive_char_sequence[-1]['s2:end'] == symbol_node['s2:start']:
-                new_symbol = _merge_consecutive_symbols(consecutive_char_sequence, consecutive_char_indices, soup.new_tag(symbol_node.name))
-                node_symbols.append(new_symbol)
-                consecutive_char_sequence = []
-                consecutive_char_indices = []
-            consecutive_char_sequence.append(symbol_node)
-            consecutive_char_indices.extend(characters)
+
+    # Parse the MathML equation, extracting top-level symbols.
+    top_level_symbols = parse_element(soup.body).symbols
+
+    # Build a list of all symbols found with a breadth-first search over the symbol tree.
+    all_symbols = []
+    symbols_to_visit = top_level_symbols
+    while len(symbols_to_visit) > 0:
+        symbol = symbols_to_visit.pop(0)
+        all_symbols.append(symbol)
+        symbols_to_visit.extend(symbol.children)
+
+    return all_symbols
+
+
+@dataclass(frozen=True)
+class ParseResult:
+
+    symbols: List[Node]
+    """
+    Top-level symbol nodes found by parsing an element. These are nodes that can be considered
+    "child symbols", if you are parsing the child tag of the parent.
+    """
+
+    element: Optional[Tag]
+    """
+    A cleaned up BeautifulSoup element of the parsed entity (i.e. with S2 metadata tags removed
+    and with consecutive identifier characters merged.). Will be None if no element could be
+    parsed from the input element.
+    """
+
+    tokens: List[Token]
+    """
+    List of all tokens found in the subtree starting at the parsed element.
+    """
+
+
+def parse_element(element: Tag) -> ParseResult:
+    """
+    Extract symbol nodes from an element in a BeautifulSoup MathML parse tree. This function
+    recursively visits
+    """
+
+    # Detect if this tag represents a KaTeX parse error. If so, return nothing.
+    if _is_error_element(element):
+        return ParseResult([], None, [])
+
+    # If this is an 'mrow' produced by KaTeX, its children are probably needlessly fragmented. For
+    # instance, the word 'true' will contain four '<mi>' elements, one for 't', 'r', 'u', and 'e'
+    # each. Merge such elements into single elements.
+    soup_children = element.children
+    if element.name == "mrow":
+        merged_children = MathMlElementMerger().merge(soup_children)
```

Comment:
curious why not make this a static method? Instead of spinning up a new instance on every merge. Or alternatively have the `soup_children` be passed in the constructor since the name implies that it will be merging some MathML elements

Reply:
I agree, a method feels like the right call here! I'll refactor this.